File: auto_editor/preview.py
from __future__ import annotations

import sys
from fractions import Fraction
from statistics import fmean, median
from typing import TextIO

from auto_editor.analyze import initLevels
from auto_editor.timeline import v3
from auto_editor.utils.bar import initBar
from auto_editor.utils.func import to_timecode
from auto_editor.utils.log import Log


def time_frame(
    fp: TextIO, title: str, ticks: float, tb: Fraction, per: str | None = None
) -> None:
    tc = to_timecode(ticks / tb, "ass")

    tp = 9 if tc.startswith("-") else 10
    tcp = 12 if tc.startswith("-") else 11
    preci = 0 if int(ticks) == ticks else 2
    end = "" if per is None else f" {per:>7}"

    fp.write(f" - {f'{title}:':<{tp}} {tc:<{tcp}} {f'({ticks:.{preci}f})':<6}{end}\n")


def all_cuts(tl: v3, in_len: int) -> list[int]:
    # Calculate cuts
    tb = tl.tb
    clip_spans: list[tuple[int, int]] = []

    for clip in tl.a[0]:
        old_offset = clip.offset * clip.speed
        clip_spans.append((round(old_offset), round(old_offset + clip.dur)))

    cut_lens = []
    i = 0
    while i < len(clip_spans) - 1:
        if i == 0 and clip_spans[i][0] != 0:
            cut_lens.append(clip_spans[i][0])

        cut_lens.append(clip_spans[i + 1][0] - clip_spans[i][1])
        i += 1

    if len(clip_spans) > 0 and clip_spans[-1][1] < round(in_len / tb):
        cut_lens.append(in_len - clip_spans[-1][1])

    return cut_lens


def preview(tl: v3, log: Log) -> None:
    log.conwrite("")
    tb = tl.tb

    # Calculate input videos length
    in_len = 0
    bar = initBar("none")
    for src in tl.unique_sources():
        in_len += initLevels(src, tb, bar, False, log).media_length

    out_len = len(tl)
    diff = out_len - in_len

    fp = sys.stdout
    fp.write("\nlength:\n")
    time_frame(fp, "input", in_len, tb, "100.0%")
    time_frame(fp, "output", out_len, tb, f"{round((out_len / in_len) * 100, 2)}%")
    time_frame(fp, "diff", diff, tb, f"{round((diff / in_len) * 100, 2)}%")

    clip_lens = [clip.dur for clip in tl.a[0]]
    log.debug(clip_lens)

    fp.write(f"clips:\n - amount:    {len(clip_lens)}\n")
    if len(clip_lens) > 0:
        time_frame(fp, "smallest", min(clip_lens), tb)
        time_frame(fp, "largest", max(clip_lens), tb)
    if len(clip_lens) > 1:
        time_frame(fp, "median", median(clip_lens), tb)
        time_frame(fp, "average", fmean(clip_lens), tb)

    cut_lens = all_cuts(tl, in_len)
    log.debug(cut_lens)
    fp.write(f"cuts:\n - amount:    {len(cut_lens)}\n")
    if len(cut_lens) > 0:
        time_frame(fp, "smallest", min(cut_lens), tb)
        time_frame(fp, "largest", max(cut_lens), tb)
    if len(cut_lens) > 1:
        time_frame(fp, "median", median(cut_lens), tb)
        time_frame(fp, "average", fmean(cut_lens), tb)

    fp.write("\n")
    fp.flush()

File: src/preview.nim
import std/sets
import std/[strutils, strformat]
import std/algorithm
from std/math import round

import av
import ffmpeg
import timeline
import log

proc timeFrame(title: string, ticks: int, tb: float, per: string = ""): string =
  let tc = toTimecode(ticks.float64 / tb, Code.ass)
  let tp = (if tc.startsWith("-"): 9 else: 10)
  let tcp = (if tc.startsWith("-"): 12 else: 11)
  let endStr = (if per == "": "" else: " " & alignLeft(per, 7))

  let titlePart = alignLeft(title & ":", tp)
  let tcPart = alignLeft(tc, tcp)
  let ticksPart = alignLeft(fmt"({ticks})", 6)

  return fmt" - {titlePart} {tcPart} {ticksPart}{endStr}"

proc timeFrame(title: string, ticks: float, tb: float, per: string = ""): string =
  let tc = toTimecode(ticks / tb, Code.ass)
  let tp = (if tc.startsWith("-"): 9 else: 10)
  let tcp = (if tc.startsWith("-"): 12 else: 11)
  let endStr = (if per == "": "" else: " " & alignLeft(per, 7))

  let titlePart = alignLeft(title & ":", tp)
  let tcPart = alignLeft(tc, tcp)
  let ticksPart = alignLeft(fmt"({ticks:.2f})", 6)

  return fmt" - {titlePart} {tcPart} {ticksPart}{endStr}"


func mean(data: seq[int]): float =
  var sum = 0
  for d in data:
    sum += d

  return sum / data.len

func median(data: seq[int]): float =
  if data.len == 0:
    return 0.0

  var sortedData = data
  sortedData.sort()

  let n = sortedData.len
  if n mod 2 == 1:
    return float(sortedData[n div 2])
  else:
    let mid1 = sortedData[(n div 2) - 1]
    let mid2 = sortedData[n div 2]
    return (mid1 + mid2) / 2


func round(a: float): int =
  int(math.round(a))

func allCuts(tl: v3, inLen: int): seq[int] =
  # Calculate cuts
  let tb = tl.tb
  var clipSpans: seq[(int, int)] = @[]

  for clip in tl.a[0].clips:
    let oldOffset = clip.offset.float64 * clip.speed
    clipSpans.add((round(oldOffset), round(oldOffset + clip.dur.float64)))

  var cutLens: seq[int] = @[]
  var i = 0
  while i < len(clipSpans) - 1:
    if i == 0 and clipSpans[i][0] != 0:
      cutLens.add(clipSpans[i][0])

    cutLens.add(clipSpans[i + 1][0] - clipSpans[i][1])
    i += 1

  if clipSpans.len > 0 and clipSpans[^1][1] < round(inLen / tb):
    cutLens.add(inLen - clipSpans[^1][1])

  return cutLens


proc preview*(tl: v3) =
  conwrite("")

  var inputLength = 0
  for src in tl.uniqueSources:
    let container = av.open(src[])
    let mediaLength: AVRational = container.mediaLength()
    inputLength += round((mediaLength * tl.tb).float64).int

  let outputLength = tl.len
  let diff = outputLength - inputLength
  let tb = tl.tb.float64

  stdout.write("\nlength:\n")
  echo timeFrame("input", inputLength, tb, "100.0%")

  if inputLength != 0:
    let outputPercent = fmt"{round((outputLength / inputLength) * 100, 2)}%"
    echo timeFrame("output", outputLength, tb, outputPercent)
    echo timeFrame("diff", diff, tb, fmt"{round((diff / inputLength) * 100, 2)}%")
  else:
    echo timeFrame("output", outputLength, tb, "0.0%")
    echo timeFrame("diff", diff, tb, "0.0%")

  var clipLens: seq[int] = @[]
  for clip in tl.a[0].clips:
    clipLens.add clip.dur

  stdout.write("clips:\n - amount:    " & $clipLens.len & "\n")
  if clipLens.len > 0:
    echo timeFrame("smallest", min(clipLens), tb)
    echo timeFrame("largest", max(clipLens), tb)
  if clipLens.len > 1:
    echo timeFrame("median", median(clipLens), tb)
    echo timeFrame("average", mean(clipLens), tb)

  let cutLens = allCuts(tl, inputLength)
  stdout.write("cuts:\n - amount:    " & $cutLens.len & "\n")
  if len(cutLens) > 0:
    echo timeFrame("smallest", min(cutLens), tb)
    echo timeFrame("largest", max(cutLens), tb)
  if len(cutLens) > 1:
    echo timeFrame("median", median(cutLens), tb)
    echo timeFrame("average", mean(cutLens), tb)

  stdout.write("\n")
  stdout.flushFile()
